gpii/node_modules/lifecycleManager/src/UserLogonRequest.js

```diff
@@ -0,0 +1,211 @@
+    logoutNoUserPromise.then(function () {
+        that.handleRequest("login");
+    }, function (error) {
+        lifecycleManager.errorResponse(that, "Error occurred during logout: " + error.message);
+    });
+};
+
+// A variant of user logon request to handle key-out requests.
+fluid.defaults("gpii.lifecycleManager.logoutRequest", {
+    gradeNames: [
+        "gpii.lifecycleManager.userLogonRequest",
+        "gpii.lifecycleManager.userLogonHandling.stateChangeHandler"
+    ],
+    listeners: {
+        "onCreate.logoutUser": {
+            listener: "{that}.handleRequest",
+            args: ["logout"]
+        }
+    }
+});
+
+/*
+ * The lifecycleManager logon request queue is used to hold the key-in and key-out requests.
+ * The entries in the action queue are of the format:
+ * {
+ *     gpiiKey: {String},
+ *     logonState: {String}  // enum: ["login", "logout"]
+ * }
+ * The action queue is run sequentially, and an item is considered "done" once the promise returned by the request
+ * is resolved.
+ */
+gpii.lifecycleManager.addToUserLogonRequestQueue = function (that, queue, item) {
+    var newItem = fluid.copy(item);
+    newItem.requestPromise = fluid.promise();
+    queue.push(newItem);
+    if (queue.length === 1) {
+        that.processUserLogonRequestQueue();
+    }
+    return newItem.requestPromise;
+};
+
+gpii.lifecycleManager.processUserLogonRequestQueue = function (that, queue) {
+    // Pick the first item and process. This item will be removed from the queue once the processing completes.
+    var item = queue[0];
+    var gradeName = gpii.lifecycleManager.userLogonRequest.handlerGrades[item.logonState];
+
+    if (!gradeName) {
+        fluid.fail("No handler grade found for user logon action - " + item.logonState);
+    };
+
+    that.events.onUserLogonRequestReceived.fire(gradeName, item.gpiiKey, item.requestPromise);
+
+    item.requestPromise.then(that.endProcessOneUserLogonRequest, that.endProcessOneUserLogonRequest);
+};
+
+gpii.lifecycleManager.endProcessOneUserLogonRequest = function (lifecycleManager) {
+    var queue = lifecycleManager.userLogonRequestQueue;
+
+    // The processing on the first item completes. Remove it from the queue.
+    queue.shift();
+
+    if (queue.length > 0) {
+        // process next item in the queue if it exists
+        lifecycleManager.processUserLogonRequestQueue(queue);
+    } else {
+        // if no actual user is logged in, login with "noUser"
+
+        // Use setTimeout() to work around the issue that kettle reports the error "Failing to resume callback for request xxx
+        // which has already concluded". The request that fails to resume callback is the user logon request that the current
+        // "endProcessOneUserLogonRequest" callback is attached with. It appears that starting the process of "noUser" keyin
+        // before the current callback completes will trigger this error.
+        setTimeout(function () {
```

Comment:
We tend to use fluid.invokeLater for such a task, but we should still try to understand what is happening here - mark with a TODO at least

Reply:
Mark with TODO to investigate reason for request failure and change setTimeout to fluid.invokeLater